src/java/org/apache/cassandra/db/memtable/AbstractAllocatorMemtable.java

```diff
@@ -0,0 +1,268 @@
+        {
+        case unslabbed_heap_buffers:
+            logger.debug("Memtables allocating with on-heap buffers");
+            return new HeapPool(heapLimit, memtableCleanupThreshold, cleaner);
+        case heap_buffers:
+            logger.debug("Memtables allocating with on-heap slabs");
+            return new SlabPool(heapLimit, 0, memtableCleanupThreshold, cleaner);
+        case offheap_buffers:
+            logger.debug("Memtables allocating with off-heap buffers");
+            return new SlabPool(heapLimit, offHeapLimit, memtableCleanupThreshold, cleaner);
+        case offheap_objects:
+            logger.debug("Memtables allocating with off-heap objects");
+            return new NativePool(heapLimit, offHeapLimit, memtableCleanupThreshold, cleaner);
+        default:
+            throw new AssertionError();
+        }
+    }
+
+    // only to be used by init(), to setup the very first memtable for the cfs
+    public AbstractAllocatorMemtable(AtomicReference<CommitLogPosition> commitLogLowerBound, TableMetadataRef metadataRef, Owner owner)
+    {
+        super(metadataRef);
+        this.commitLogLowerBound = commitLogLowerBound;
+        this.allocator = MEMORY_POOL.newAllocator();
+        this.initialComparator = metadata.get().comparator;
+        this.owner = owner;
+        scheduleFlush();
+    }
+
+    public CommitLogPosition getApproximateCommitLogLowerBound()
+    {
+        return approximateCommitLogLowerBound;
+    }
+
+    public MemtableAllocator getAllocator()
+    {
+        return allocator;
+    }
+
+    public boolean shouldSwitch(ColumnFamilyStore.FlushReason reason)
+    {
+        switch (reason)
+        {
+        case SCHEMA_CHANGE:
+            return initialComparator != metadata().comparator // If the CF comparator has changed, because our partitions reference the old one
+                   || metadata().params.memtable.factory != factory(); // If a different type of memtable is requested
+        default:
+            return true;
+        }
+    }
+
+    public void metadataUpdated()
+    {
+        scheduleFlush();
+    }
+
+    public void performSnapshot(String snapshotName)
+    {
+        // unless shouldSwitch(SNAPSHOT) returns false, this cannot be called.
+        throw new AssertionError();
+    }
+
+    protected abstract Factory factory();
+
+    public void switchOut(OpOrder.Barrier writeBarrier, AtomicReference<CommitLogPosition> commitLogUpperBound)
+    {
+        super.switchOut(writeBarrier, commitLogUpperBound);
+        this.commitLogUpperBound = commitLogUpperBound;
+        allocator.setDiscarding();
+    }
+
+    public void discard()
+    {
+        allocator.setDiscarded();
+    }
+
+    // decide if this memtable should take the write, or if it should go to the next memtable
+    public boolean accepts(OpOrder.Group opGroup, CommitLogPosition commitLogPosition)
+    {
+        // if the barrier hasn't been set yet, then this memtable is still taking ALL writes
```

Comment:
> // if the barrier hasn't been set yet, then this memtable is still taking ALL writes
I find this comment misleading.

If the barrier hasn't been set yet then the write must have been created with such opGroup and such commitLogPosition that both an issued barrier would permit it and `LastCommitLogPosition >= commitLogPosition`. That's because setting the barrier happens before opGroup switch and fixing `LastCommitLogPosition`. 

This is why we do not care about the race condition in:
```
        if (barrier == null)
            return true;
```

does ^ make sense? Could we change the comment?

Reply:
I also tend to keep forgetting why all this is necessary. Expanded the comment with further details.

Let me know if this is now clear enough.